engine/engine.go

```diff
@@ -128,6 +132,14 @@ func New(opts Opts) *compatibilityEngine {
 		opts.LookbackDelta = 5 * time.Minute
 		level.Debug(opts.Logger).Log("msg", "lookback delta is zero, setting to default value", "value", 5*time.Minute)
 	}
+	if opts.ExtLookbackDelta == 0 {
+		opts.ExtLookbackDelta = 1 * 24 * time.Hour
+	}
+
+	// Set the parser functions for extended functions appropriately as they are not present in prometheus
+	parser.Functions["xdelta"] = parse.Functions["xdelta"]
+	parser.Functions["xincrease"] = parse.Functions["xincrease"]
+	parser.Functions["xrate"] = parse.Functions["xrate"]
```

Comment:
Maybe we want this in an `init` function so we don't have to do it for each instance of the engine.

Reply:
We were planning on making it possible to switch between engines using headers to improve the UX i.e. so that a whole restart wouldn't be needed to switch between engines and it would allow faster experimentation with engines. I believe this will interfere with how the Prometheus engine works? :cry: we will probably have to fork the parser :/

Reply:
What if we have a custom implementation of `ParseExpr` that injects the new functions, calls Prometheus `ParseExpr` and removes the functions once it is done? We might need to use a mutex but I don't think the impact should be that large in comparison to the cost of executing the query.

Maintaining a fork of the Prometheus parser sound scary 😀

Reply:
That sounds very hacky :smile: I mean we wouldn't even have to change anything in the parser, the only reason we'd have that fork is to avoid clashing between engines so it should be very easy to update. Perhaps we could just vendor it inside `vendor/` in this repo with some kind of different name and then use it in `go.mod`?

Reply:
Thinking about this again, I think vendoring might be the better option. I am concerned we could push ourselves into a corner if we fork Prometheus since we will import two modules with the same name `github.com/prometheus/prometheus`.  

I don't know if it's a valid concern, but maybe it's better to vendor the parser just to be on the safe side.

Reply:
Is there a possibility to vendor only one module? `go mod vendor` seems to suggest that it's all or nothing :sweat_smile:

Reply:
Something like what we do for Cortex in Thanos: https://github.com/thanos-io/thanos/pull/5504?

Reply:
I think the decision to vendor Cortex under an internal package was because of circular dependencies.

I'll need to check what happens if we vendor prometheus using `go mod`. My concern is that Thanos will require prometheus/prometheus from github.com, and promql-engine will require prometheus/prometheus from a local vendor folder. 

In the ideal case this will not cause Thanos to somehow magically use prometheus/prometheus from promql's vendored folder, but I would like to double check that first.

Reply:
We tried to vendor `prometheus/prometheus` within `promql-engine` but go module system starts running into issues when you reference both `promql-engine` and a separate `prometheus/prometheus` within a third project.

So for example, if we import `thanos-community/promql-engine` and a separate `prometheus/prometheus` within a third project it will just optimise the vendored Prometheus away.

We tried to force it to use separate versions of Prometheus to see if it would avoid the optimisation but it simply pruned the whole `thanos-community/promql-engine` module once it detected a version mismatch of prometheus between the two.

Reply:
And another update on this, I tried to simply copy the parser code using a make target into the engine but we run into some more issues with compatibility.

For example the Result expects the prom parser to be defined there: https://github.com/prometheus/prometheus/blob/v0.42.0/promql/value.go#L296

![Screenshot 2023-03-21 at 11 22 40](https://user-images.githubusercontent.com/87335168/226694444-fe1cee64-3dbb-4dc2-a8b0-3fd53c53505c.png)

Reply:
Mhm, so it seems like a full fork of the repository is the only solution? Thanks for trying out all of these options 🙇

Reply:
Or we just inject the functions as initially implemented 😞 but I realise that leads to some undesirable side effects.

Reply:
We can try the fork, but I suspect it will have the same problem as the vendoring solution. The fork's module name will be `prometheus/prometheus` so when we do a require from Thanos, the `go mod` command will have to converge a single module. My guess is that we won't be able to `require` both our fork and github.com/prometheus/prometheus from inside Thanos, even though the fork is a transitive dependency.

Reply:
Unfortunately I think you're correct, I tried forking [prometheus](https://github.com/mklocke-shopify/prometheus/commit/f2d98bc2f97bfb0c09bb40c499e7da850d6c9054) and then replaced in promql-engine:

`replace github.com/prometheus/prometheus v0.42.0 => github.com/mklocke-shopify/prometheus v0.42.1-0.20230321212756-f2d98bc2f97b`

When both `promql-engine` and `prometheus` are imported into a third go application it optimises the imports in such a manner that my fork in `promql-engine` was ignored and the functions were no longer included in the parser.

Reply:
Why not replace the whole import so the module's path will be `github.com/mklocke-shopify/prometheus` instead of `github.com/prometheus/prometheus`?

Reply:
Unfortunately this is currently not possible with go modules because the module name is defined in the `go.mod` [file](https://github.com/grafana/mimir/blob/main/go.mod#L245-L246). If we change the module name of the fork, we also have to modify all internal imports that point to `github.com/prometheus/prometheus` to point to `github.com/thanos-io/prometheus`.

Mimir also has a Prometheus fork and they point to it using a replace statement: https://github.com/grafana/mimir/blob/main/go.mod#L245-L246. 

Maybe we should address the root cause here is that the parser is not configurable and relies on the global `Functions` variable. I've raised a PR to address this: https://github.com/prometheus/prometheus/pull/12170.

Reply:
It seems that Prometheus might in fact be willing to take in a contribution for xrate/xincrease, but Julien is reluctant to change the parser to allow injecting additional functions: https://github.com/prometheus/prometheus/pull/12170#issuecomment-1479320537. 

Any suggestions on what else we can do here?

Reply:
If they are considering accepting xFunctions behind a feature flag some day (probably not soon?) can we accept this as is with injecting it into the parser? I realise this causes potential unwanted side-affects when importing both this and original prometheus engine.

Reply:
> If we change the module name of the fork, we also have to modify all internal imports which point to github.com/prometheus/prometheus to point to github.com/thanos-io/prometheus.

Why isn't this an option? 🤔 what are the blockers?

Reply:
What I think we could do is to have separate go module for https://github.com/prometheus/prometheus/promql/parser part. Or we extract that part of code to a new repo and we make changes.
Cortex uses both Prometheus and promql-engine, but we don't want to use Prometheus fork.

Reply:
> Why isn't this an option? 🤔 what are the blockers?

I would be concerned with the maintenance burden. Once we change all internal references in the fork, we won't be able to update it through the UI and we will need some automation to pull latest changes from Prometheus and also resolve conflicts with imports. It might be doable but I am concerned it can get quite error prone.

However, this could be a way to vendor in the Prometheus parser: https://github.com/thanos-community/promql-engine/pull/223.